ARCHITECTURE.md

```diff
@@ -0,0 +1,170 @@
+
+### Architecture Diagram
+
+![kURL sh Architecture](https://user-images.githubusercontent.com/371319/166266866-267833a7-b21f-4665-b657-a55d6271b48b.png)
+
+### Object Storage
+
+| Object | Description |
+| ------ | ----------- |
+| s3://[bucket]/[kURL version]/[entrypoint].tmpl           | Script templates rendered by the API. install.sh, join.sh, upgrade.sh and tasks.sh |
+| s3://[bucket]/[kURL version]/[addon-version].tar.gz      | Add-on bundles |
+| s3://[bucket]/[kURL version]/supported-versions-gen.json | Add-on and version information for a given release of kURL. For resolving "latest" and "dot x" versions as well as validating installers. |
+
+## kURL Installer
+
+### Entrypoints
+
+| Script | Description |
+| ------ | ----------- |
+| install.sh | Initialize and upgrade a kURL cluster. |
+| join.sh    | Join additional nodes to the cluster. |
+| upgrade.sh | Utility script for upgrading additional nodes. |
+| tasks.sh   | For various tasks including re-generating the join script and loading images from an air-gap bundle. |
+
+### kURL Directory Structure
+
+Exists on the server at `/var/lib/kurl`.
+
+| Directory | Description |
+| ----------| ----------- |
+| addons/    | Stores addons, including images and assets. |
+| bin/       | Stores utility binaries. |
+| krew/      | Has some plugins we use including support-bundle and preflight. |
+| kurlkinds/ | Contains cluster.kurl.sh CRD |
+| kustomize/ | Scripts output  |
+| packages/  | Host packages |
+| shared/    | Everything else |
+
+### Flow Chart
+
+#### Installation Lifecycle
+
+![kURL sh Installation Lifecycle](https://user-images.githubusercontent.com/371319/166290254-f73a4232-5227-45b5-a673-cc742e626e57.png)
+
+#### Join Lifecycle
+
+![kURL sh Join Lifecycle](https://user-images.githubusercontent.com/371319/166290269-b6451622-2890-491e-98cc-6f20997c8b13.png)
+
+## Add-ons
+
+Add-ons are components that make up a kURL cluster.
+
+### Categories
+
+1. Kubernetes distribution - Kubeadm, RKE2, K3s
+1. CRI - Docker or Containerd
+1. CNI - Weave or Antrea
+1. CSI - Longhorn, Rook, OpenEBS
+1. Ingress - Contour
+1. Misc. - Prometheus, KOTS, Velero...
+
+### Directory Structure
+
+| Directory | Description |
+| --------- | ----------- |
+| addons/[addon]/[version]/Manifest            | Manifest of assets, host packages and container images |
+| addons/[addon]/[version]/install.sh          | Entrypoint to the add-on installation script |
+| addons/[addon]/[version]/host-preflight.yaml | Troubleshoot.sh preflight spec |
+| addons/[addon]/[version]/assets/             | Runtime assets |
+| addons/[addon]/[version]/images/             | Runtime images |
+| addons/[addon]/[version]/[distro-version]/   | Runtime host packages for each supported Linux OS  |
+
+### Lifecycle
+
+#### addon_fetch
+
+Fetch the add-on package from the object store or from the air-gap bundle.
+This step is typically skipped if the add-on version has not changed since the previous run.
```

Comment:
I'd mention that we're fetching this _to `/var/lib/kurl/addons`_, because otherwise it's a little confusing